tsdb/engine/tsm1/engine.go

```diff
@@ -2157,30 +2165,36 @@ func (e *Engine) compact(wg *sync.WaitGroup) {
 			if level, runnable := e.scheduler.next(); runnable {
 				switch level {
 				case 1:
-					if e.compactHiPriorityLevel(level1Groups[0], 1, false, wg) {
+					if e.compactHiPriorityLevel(level1Groups[0], 1, false, wg, tsdb.DefaultMaxPointsPerBlock) {
 						level1Groups = level1Groups[1:]
 					}
 				case 2:
-					if e.compactHiPriorityLevel(level2Groups[0], 2, false, wg) {
+					if e.compactHiPriorityLevel(level2Groups[0], 2, false, wg, tsdb.DefaultMaxPointsPerBlock) {
 						level2Groups = level2Groups[1:]
 					}
 				case 3:
-					if e.compactLoPriorityLevel(level3Groups[0], 3, true, wg) {
+					if e.compactLoPriorityLevel(level3Groups[0], 3, true, wg, tsdb.DefaultMaxPointsPerBlock) {
 						level3Groups = level3Groups[1:]
 					}
 				case 4:
+					var pointsPerBlock int
 					// This is a heuristic. The 10_000 points per block default is suitable for when we have a
 					// single generation with multiple files at max block size under 2 GB.
 					if genLen == 1 {
 						// Log TSM files that will have an increased points per block count.
 						for _, f := range level4Groups[0] {
 							e.logger.Info("TSM optimized compaction on single generation running, increasing total points per block.", zap.String("path", f), zap.Int("points-per-block", e.CompactionPlan.GetAggressiveCompactionPointsPerBlock()))
 						}
-						e.Compactor.Size = e.CompactionPlan.GetAggressiveCompactionPointsPerBlock()
+						pointsPerBlock = tsdb.DefaultAggressiveMaxPointsPerBlock
 					} else {
-						e.Compactor.Size = tsdb.DefaultMaxPointsPerBlock
+						pointsPerBlock = tsdb.DefaultMaxPointsPerBlock
+						for _, f := range level4Groups[0] {
+							if tsmPointsPerBlock := e.Compactor.FileStore.BlockCount(f, 1); tsmPointsPerBlock == tsdb.DefaultAggressiveMaxPointsPerBlock {
+								pointsPerBlock = tsdb.DefaultAggressiveMaxPointsPerBlock
+							}
```

Comment:
We can break out of the loop once we decide we need to stick with the aggressive points per block to avoid going through the rest of the TSM files.